### PR TITLE
sharding: balance shards waiting lists with shuffled out nodes

### PR DESCRIPTION
**Pull request by AdoAdoAdo**
with activation of capped shuffling of nodes, the shards waiting lists may become unbalanced
this may cause waiting times to differ between shards waiting lists
this PR will try at the end of each epoch to first equalize the waiting lists.

### CODE REVIEW

File: sharding/hashValidatorShuffler.go [hunk 1]
```diff
@@ -657,6 +706,9 @@ func (rhs *randHashShuffler) UpdateShufflerConfig(epoch uint32) {
 		"epochEnable", rhs.activeNodesConfig.EpochEnable,
 		"maxNodesToShufflePerShard", rhs.activeNodesConfig.NodesToShufflePerShard,
 	)
+
+	rhs.flagBalanceWaitingLists.Toggle(epoch >= rhs.balanceWaitingListsEnableEpoch)
+	log.Trace("balanced waiting lists", "enabled", rhs.flagBalanceWaitingLists.IsSet())
```
Comment (iulianpascalau) [inline sharding/hashValidatorShuffler.go:711]: log.Debug as to align with the rest of the project.

File: sharding/hashValidatorShuffler.go [hunk 2]
```diff
@@ -250,12 +257,14 @@ func shuffleNodes(arg shuffleNodesArg) (*ResUpdateNodes, error) {
 	if err != nil {
 		log.Warn("moveNodesToMap failed", "error", err)
 	}
-	err = distributeValidators(newWaiting, arg.newNodes, arg.randomness)
+
+	err = distributeValidators(newWaiting, arg.newNodes, arg.randomness, false)
+
```
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply sharding/hashValidatorShuffler.go]: done

File: sharding/hashValidatorShuffler.go (hunk 1 above)
Comment (AdoAdoAdo) [reply sharding/hashValidatorShuffler.go]: done

**Review by iulianpascalau — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed

**Pull request merged**